src/sc_client/core/async_sc_connection.py

```diff
@@ -0,0 +1,156 @@
+            self._responses_dict[response.id] = response
+
+    async def _on_event(self, response: Response):
+        event = self.get_event(response.id)
+        if event is None:
+            raise ScEventError(ErrorNotes.EVENT_IS_NOT_FOUND, response.id)
+        self._logger.debug("Started %s", str(event))
+        addrs: list[ScAddr] = [ScAddr(value) for value in response.payload]
+        asyncio.create_task(event.callback(*addrs), name=f"ScEvent({event.id})")
+        await asyncio.sleep(0)
+
+    def set_event(self, sc_event: AsyncScEvent) -> None:
+        self._events_dict[sc_event.id] = sc_event
+
+    def get_event(self, event_id: int) -> AsyncScEvent | None:
+        return self._events_dict.get(event_id)
+
+    def drop_event(self, event_id: int) -> None:
+        del self._events_dict[event_id]
+
+    async def disconnect(self) -> None:
+        if self.is_connected():
+            await self._websocket.close()
+            self._logger.info("Connection closed")
+            await self.on_close()
+        else:
+            self._logger.info("Connection was already closed")
+
+    def is_connected(self) -> bool:
+        return self._websocket.open
+
+    async def send_message(self, request_type: common.RequestType, payload: any) -> Response:
+        self._command_id += 1
+        command_id = self._command_id
+        data = json.dumps(
+            {
+                common.ID: command_id,
+                common.TYPE: request_type.value,
+                common.PAYLOAD: payload,
+            }
+        )
+        len_data = len(bytes(data, "utf-8"))
+        if len_data > config.MAX_PAYLOAD_SIZE:
+            raise PayloadMaxSizeError(f"Data is too large: {len_data} > {config.MAX_PAYLOAD_SIZE} bytes")
+        await self._send_with_reconnect(data)
+        response = await self._receive(command_id)
+        return response
+
+    async def _send_with_reconnect(self, data: str) -> None:
+        retries: int = self.reconnect_retries
+        while True:
+            try:
+                await self._websocket.send(data)
+                return
+            except ConnectionClosedOK as e:
+                raise ScServerError(ErrorNotes.ALREADY_DISCONNECTED) from e
+            except ConnectionClosed as e:
+                if retries <= 0:
+                    raise ScServerError(ErrorNotes.SC_SERVER_TAKES_A_LONG_TIME_TO_RESPOND) from e
+                retries -= 1
+                self._logger.warning(
+                    "Trying to reconnect in %d seconds (retries left: %s)", self.reconnect_delay, retries
+                )
+                await asyncio.sleep(self.reconnect_delay)
+                try:
+                    await self.connect()
+                    await self.on_reconnect()
+                except ScServerError:
+                    pass
+
+    async def _receive(self, command_id: int) -> Response:
+        while (answer := self._responses_dict.get(command_id)) is None and self._websocket.open:
```

Comment:
This implementation does not fix the exist problems. Instead of just sending a message, attaching a function to the message ID that will be called after receiving a response, we constantly wait for a response to the message, checking this in a loop. It's very expensive. I would suggest returning a promizу object, as is done in ts-sc-client, and remembering the message ID of the message response processing function. The cycle needs to be removed here.

Reply:
I replaced it using `asyncio.Future`. There is no cicle now